### PR TITLE
Improve the test case for exceeding the headroom limit

### PR DESCRIPTION
**Pull request by stephenxs**
- Don't use hard-coded value of the super long cable length
  Find out the longest cable length by retrying in a loop
- Use log analyzer to verify whether the expected error messages are found

Signed-off-by: Stephen Sun <stephens@nvidia.com>

### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### CODE REVIEW

File: tests/qos/test_buffer.py [hunk 1]
```diff
@@ -1228,54 +1264,146 @@ def test_exceeding_headroom(duthosts, rand_one_dut_hostname, conn_graph_facts, p
...
+                                                 'doTask: Failed to process buffer task, drop it'])
+        logging.info('[Find out the longest cable length the port can support]')
+        cable_length = 300
+        while True:
+            duthost.shell('config interface cable-length {} {}m'.format(port_to_test, cable_length))
+            expected_profile = 'pg_lossless_{}_{}m_profile'.format(original_speed, cable_length)
+            profile_applied = check_pg_profile(duthost, 'BUFFER_PG_TABLE:{}:3-4'.format(port_to_test), expected_profile, False)
+            if not profile_applied:
+                break
+            logging.info('Cable length {} has been applied successfully'.format(cable_length))
```
Comment (keboliu) [inline tests/qos/test_buffer.py:1295]: should we only log when find out the longest cable length?

File: tests/qos/test_buffer.py [hunk 2]
```diff
@@ -1228,54 +1264,146 @@ def test_exceeding_headroom(duthosts, rand_one_dut_hostname, conn_graph_facts, p
...
+            cable_length += 100
+
+        # We've got the maximum cable length that can be applied on the port
+        violating_cable_length = cable_length
+        maximum_cable_length = cable_length - 100
```
Comment (keboliu) [inline tests/qos/test_buffer.py:1300]: can we have these two variables assigned in the loop?

File: tests/qos/test_buffer.py (hunk 1 above)
Comment (stephenxs) [reply tests/qos/test_buffer.py:1295]: What happens if the maximum headroom is very large? It takes ~5 seconds each iteration. To print something every iteration let user knows it is running.

File: tests/qos/test_buffer.py (hunk 2 above)
Comment (stephenxs) [reply tests/qos/test_buffer.py:1300]: Yes, we can if we think that’s a better way. But I don’t see any benefits?

**Review by keboliu — APPROVED** [review]

**Comment by stephenxs** [comment]
Community PR [3062](https://github.com/Azure/sonic-mgmt/pull/3062) opened

**Pull request closed without merging**